### PR TITLE
Add outputs tags function

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
The major goal of this is to add `add_output_tags` to the step context.
Side effect: Artifacts also are taggable via create call.

OSSK-389

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/user-guide/starter-guide/manage-artifacts.md [hunk 1]
```diff
@@ -121,6 +121,66 @@ The Cloud dashboard visualizes version history for your review.
...
+If you would like to extend your Artifacts with extra metadata or tags you can do it following patterns demonstrated below:
+
+```python
+from zenml import step, get_step_context, ArtifactConfig
+from typing_extensions import Annotated
+
+
+# below we annotate output with `ArtifactConfig` giving it a name,
+# run_metadata and tags. As a result, created artifact `artifact_name`
+# will get configured metadata and tags
```
Comment (strickvl) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:135]:
```suggestion
# run_metadata and tags. As a result, the created artifact
# `artifact_name` will get configured with metadata and tags
```

File: docs/book/user-guide/starter-guide/manage-artifacts.md [hunk 2]
```diff
@@ -121,6 +121,66 @@ The Cloud dashboard visualizes version history for your review.
...
+            tags=["tag_name"],
+        ),
+    ]
+):
+    return "string"
+
+
+# below we annotate output using functional approach with
+# run_metadata and tags. As a result, created artifact `artifact_name`
+# will get configured metadata and tags
```
Comment (strickvl) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:152]: see above

File: docs/book/user-guide/starter-guide/manage-artifacts.md [hunk 3]
```diff
@@ -121,6 +121,66 @@ The Cloud dashboard visualizes version history for your review.
...
+def annotation_approach() -> Annotated[str, "artifact_name"]:
+    step_context = get_step_context()
+    step_context.add_output_metadata(
+        output_name="artifact_name", metadata={"metadata_key": "metadata_value"}
+    )
+    step_context.add_output_tags(output_name="artifact_name", tags=["tag_name"])
+    return "string"
+
+
+# below we combine both approaches, so artifact will get
```
Comment (strickvl) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:163]:
```suggestion
# below we combine both approaches, so the artifact will get
```

File: src/zenml/models/v2/core/artifact.py [hunk 4]
```diff
@@ -44,6 +44,11 @@ class ArtifactRequest(BaseRequest):
         title="Whether the name is custom (True) or auto-generated (False).",
         default=False,
     )
+    tags: Optional[List[str]] = Field(
```
Comment (strickvl) [inline src/zenml/models/v2/core/artifact.py:47]: don't we need an alembic migration as well for this PR, given this change?

File: src/zenml/models/v2/core/artifact.py [hunk 5]
```diff
@@ -44,6 +44,11 @@ class ArtifactRequest(BaseRequest):
...
+        title="Tags of the artifact.",
```
Comment (strickvl) [inline src/zenml/models/v2/core/artifact.py:48]:
```suggestion
        title="Artifact tags.",
```

File: tests/integration/functional/steps/test_step_context.py [hunk 6]
```diff
@@ -109,3 +112,92 @@ def _pipeline():
...
+            names + data
+            for names in [
+                [
+                    "custom_name",
+                ]
+                * 2,
+                [
+                    "output_1",
+                    "_pipeline::step_context_metadata_and_tags_adder::output_1",
+                ],
```
Comment (strickvl) [inline tests/integration/functional/steps/test_step_context.py:150]: formatting seems a bit weird here?

**Review by strickvl — APPROVED** [review]

**Review by bcdurak — APPROVED** [review]
🦭

**Pull request merged**